### PR TITLE
nixos/wireguard-networkd: init

### PR DESCRIPTION
**Pull request by Majiir**
## Description of changes

Adds a networkd backend for the `networking.wireguard` options.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] armv7l-linux (cross from x86_64-linux)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - [x] `nixosTests.wireguard.*`
    - [x] `nixosTests.systemd-networkd`
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by onny** [comment]
Wow this is great, I'm really looking forward using this!

**Comment by Majiir** [comment]
Rebased, fixed a few warnings, and re-ran tests. I've also been running this on a few machines with no issues.

File: nixos/modules/services/networking/wireguard-networkd.nix [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+      '';
+    };
+  };
+
+  config = mkIf (cfg.enable && cfg.useNetworkd) {
+
+    # TODO: Some of these options may be possible to support in networkd.
+    #
+    # privateKey and presharedKey are trivial to support, but we deliberately
+    # don't to discourage putting secrets in the /nix store.
```
Comment (MinerSebas) [inline nixos/modules/services/networking/wireguard-networkd.nix:101]: Initially, I read that as a double negation as I skipped the `to`.
Hopefully, the comma should prevent this for other readers

```suggestion
    # don't, to discourage putting secrets in the /nix store.
```

Comment (Majiir) [reply nixos/modules/services/networking/wireguard-networkd.nix]: I don't think that would be grammatically correct, but I rephrased it to address the confusion:

> privateKey and presharedKey are trivial to support, but we deliberately don't **in order to** discourage putting secrets in the /nix store.

**Comment by Majiir** [comment]
- Rebased on latest `master`.
- Removed `mdDoc` usages.
- Added assertion around new `preShutdown` option.

**Review by misuzu — COMMENTED** [review]
New modules deserve a release notes entry, everything else LGTM

**Comment by Majiir** [comment]
- Rebased on latest `master`.
- Added a release notes entry.

**Comment by misuzu** [comment]
Some files need formatting to satisfy CI

**Comment by Majiir** [comment]
- Fixed formatting

**Review by misuzu — APPROVED** [review]

**Comment by Scrumplex** [comment]
This might be a breaking change for some people. In my case, networkd fails to set up the netdev because the service can't read my root-owned (mode `400`) secret:

```
wg-scrumplex: Failed to read private key from /run/agenix/wireguard.key. Ignoring network device.
```

**Comment by misuzu** [comment]
> Maybe LoadCredential should be used for stateVersion < 25.05 to ensure that the secret is readable for existing setups?

Might be easier and faster to just change the default value of `networking.wireguard.useNetworkd` for `stateVersion < 25.05`

**Comment by eliasp** [comment]
Wouldn't it make more sense in general, not only limited to `stateVersion < 25.05` to:
- configure the private key as credential for `systemd-networkd.service`, e.g. `LoadCredential=wg-${name}-private-key:${privateKeyFile}`
- point `PrivateKey=` of the WireGuard `.netdev` to the credential using the `@` prefix: `PrivateKey=@wg-${name}-private-key`

This should:
- ensure the credential's permissions don't need to be manually adjusted for `systemd-networkd`'s user to be able to read it
- work independent of the stateVersion
- use the most secure "channel" for filesystem-based service credentials there currently is, since permissions can be kept to a minimum

**Comment by Majiir** [comment]
Yes, I'll add credentials support. It wasn't in the original PR since networkd only added credential support for Wireguard recently.

I'd prefer this over a `stateVersion` check. But if this is causing trouble in the meantime, we can default `networking.wireguard.useNetworkd` to `false` until it's ready to be enabled by default.

File: nixos/modules/services/networking/wireguard-networkd.nix [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    # networkd configures interfaces. There is also a systemd feature request
+    # for key generation: https://github.com/systemd/systemd/issues/14282
+    #
+    # preSetup, postSetup, preShutdown and postShutdown may be possible, but
+    # networkd is not likely to support script hooks like this directly. See:
+    # https://github.com/systemd/systemd/issues/11629
+    #
+    # socketNamespace and interfaceNamespace can be implemented once networkd
+    # supports setting a netdev's namespace. See:
+    # https://github.com/systemd/systemd/issues/11629
```
Comment (networkException) [inline nixos/modules/services/networking/wireguard-networkd.nix:132]: I assume you meant

```suggestion
    # https://github.com/systemd/systemd/issues/11103
```

here?

**Comment by Majiir** [comment]
Quick PR to default `networking.wireguard.useNetworkd` to `false` for now: https://github.com/NixOS/nixpkgs/pull/364046

File: nixos/modules/services/networking/wireguard-networkd.nix (hunk 2 above)
Comment (Majiir) [reply nixos/modules/services/networking/wireguard-networkd.nix:132]: Oops, thanks! Fixed in https://github.com/NixOS/nixpkgs/pull/364046

**Comment by Majiir** [comment]
Switched to using credentials in https://github.com/NixOS/nixpkgs/pull/364078

**Pull request merged**